coding_standards.md

```diff
@@ -0,0 +1,95 @@
+
+### All code is checked for readability
+
+> There are only two hard things in Computer Science: cache invalidation and naming things.
+>
+> -- Phil Karlton
+
+Avoid basic errors such as
+
+- Generic variable names: `p`, `temp`, `v`
+- Using arrays over hashes: `['Bob', 'Smith', 'Bristol']` over `{ name: 'Bob', surname: 'Smith', location: 'Bristol' }`
+- Use of acronyms: particularly domain specific ones. Someone new to the service may not know their GOR from their WIMS
+
+Other than that when considering the name for a method, class, module etc try not to approach it from what makes sense to you, but instead what best describes the intent of the thing, and what will help the person who comes after you.
+
+Stick to the conventions and naming standards of the language or framework you are using
+
+Aim to write readable code that does not require comments.
+
+Use the principle of **separate in order to name**. Essentially break your code down into small (and if necessary one line) sub routines that have names that express what their intent is. Read [Write readable code without comments](http://www.wikihow.com/Write-Readable-Code-Without-Comments) for more on how to do this.
+
+If you have to add comments, make them useful
+
+- **Comments are code** - Keep them small, simple and re-factor them as you make changes
+- **Don't document the _how_** - Document *why* the thing is needed, or *why* its doing it the way it is
+- **Comment in the right place** - Talk about the module at the module declaration, the subroutine at the method declaration etc
+- **Only write about the code** - Don't include names, dates, commit comments or your opinions
+
+Don't include commented out code. It's noise that only confuses. Use the commit history if you need to see what was there before.
+
+And most importantly write them *as you write the code*. If you don't, you probably never will!
+
+### All code is appropriately documented
+
+Document the *how* for the project
+
+When it comes to other developers or users of the project, the thing they are most interested in is the *how*. *How* can I build/deploy/run this project. *How* can I use it in my own work.
+
+### All code is checked for simplicity
+
+We don't code for reuse until we have a confirmed need for reuse.
+
+We apply the [rule of three](https://en.wikipedia.org/wiki/Rule_of_three_(computer_programming)) to help decide when to abstract code.
+
+We don't break up our applications until there is a confirmed need to do so.
+
+The path to reuse starts with module/namespacing. When the need for reuse is confirmed we first build a package. Only if there is demand from multiple services, and the benefits for a deployed service over a shared package are clear do we build a new application.
+
+### All code is checked for quality
+
+All repos are connected to a quality analysis tool (such as CodeClimate or SonarQube) and the tool's maximum quality rating is maintained.
```

Comment:
Similar to my other comment, presume we don't want to retrofit this?  Suddenly applying code quality to years old legacy code would likely give us a lot of issues to fix!

Reply:
Yes, as Dave mentioned it's on a case-by-case basis.

I should possibly put this in as one of the general principles for our standards.

I'll see what more feedback I get.

Reply:
I'd add there is even more a case for checking older *legacy* code. For example in my current team we have made a point about making the metrics for the services we are working on visible. It's been fantastic in the conversations we often face about

- why it can take longer than expected to deliver features
- why the quality of our estimates may vary
- why we keep having to deal with bugs

![metrics_v3](https://user-images.githubusercontent.com/1789650/55623315-8c2f4900-579a-11e9-9d7d-16e7843ad7a4.JPG)

The difference is whether the tool **forces** you to rectify to issues. I'm also against that, but I think running the checks should be mandatory so at least the issues are visible.

Reply:
I agree. I don't think we should mandate forcing of building the issues. I do think we should be scanning for them though.